### PR TITLE
Check for snap server

### PR DESCRIPTION
**Pull request by pinges**
## PR description
When we are connecting to an EthPeer and we are doing snap sync find out whether that peer does serve snap data.
This is needed to make sure that we have a minimum of snap data serving peers and to make sure that we only send snap requests to peers that serve snap data.

### CODE REVIEW

**Review by garyschulte — COMMENTED** [review]
It looks like this is still in development, with a lot of commented code, debug and TODOs.  
The method of checking if a server is serving snap data seems solid, use their reported chain head and ensure they serve account data for it 👍 

Overall, I'd prefer that we only weight snap servers while we are in the process of snap sync however.  I think a less intrusive way to do this would be to schedule a peer check in the snap sync pipeline that stops once we have downloaded all of the snap data.  It can maintain the peer ratio and disconnect peers as needed, and we can encapsulate that behavior into just sync where it is relevant.

File: besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java [hunk 1]
```diff
@@ -668,13 +668,14 @@ public Runner build() {
             .map(nodePerms -> PeerPermissions.combine(nodePerms, bannedNodes))
             .orElse(bannedNodes);
 
+    final EthPeers ethPeers = besuController.getEthPeers();
```
[the remark this answers is not in the document]
Comment (macfarla) [reply besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java:671]: done

File: besu/VERSION_METADATA.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"besuVersion":"24.2.0-dev-5449acc7"}
```
[the remark this answers is not in the document]
Comment (macfarla) [reply besu/VERSION_METADATA.json]: done

**Comment by macfarla** [comment]
made RetryingGetAccountRangeFromPeerTask use switching peer because seeing a lot of these errors

```
{"@timestamp":"2024-04-12T07:44:00,963","level":"WARN","thread":"EthScheduler-Timer-0","class":"RequestDataStep","message":"Error handling account download accounts (0x1000000000000000000000000000000000000000000000000000000000000000 - 0x1fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) task: org.hyperledger.besu.ethereum.eth.manager.exceptions.MaxRetriesReachedException: Task failed: MAX_RETRIES_REACHED","throwable":""}
{"@timestamp":"2024-04-12T07:44:00,963","level":"WARN","thread":"EthScheduler-Timer-0","class":"RequestDataStep","message":"Error handling account download accounts (0x2000000000000000000000000000000000000000000000000000000000000000 - 0x2fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) task: org.hyperledger.besu.ethereum.eth.manager.exceptions.MaxRetriesReachedException: Task failed: MAX_RETRIES_REACHED","throwable":""}
{"@timestamp":"2024-04-12T07:44:00,963","level":"WARN","thread":"EthScheduler-Timer-0","class":"RequestDataStep","message":"Error handling account download accounts (0x3000000000000000000000000000000000000000000000000000000000000000 - 0x3fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) task: org.hyperledger.besu.ethereum.eth.manager.exceptions.MaxRetriesReachedException: Task failed: MAX_RETRIES_REACHED","throwable":""}
```

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java [hunk 3]
```diff
@@ -28,7 +28,7 @@
 import org.apache.tuweni.bytes.Bytes32;
 
 public class RetryingGetAccountRangeFromPeerTask
-    extends AbstractRetryingPeerTask<AccountRangeMessage.AccountRangeData> {
+    extends AbstractRetryingSwitchingPeerTask<AccountRangeMessage.AccountRangeData> {
```
Comment (garyschulte) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java:31]: FYI making this a peer switching task will break the snap server check logic that @pinges uses in `SnapServerChecker`

Comment (garyschulte) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/snap/RetryingGetAccountRangeFromPeerTask.java:31]: NVM - the peer checker code uses the non-retrying version, GetAccountRangeFromPeerTask

**Comment by macfarla** [comment]
with recent updates (importantly using SNAP and CHECKPOINT not X_SNAP X_CHECKPOINT), nodes are getting in sync eg 
Screenshot 2024-04-15 at 6 29 44 PM

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java [hunk 4]
```diff
@@ -695,6 +704,16 @@ public BesuController build() {
             ethProtocolManager,
             pivotBlockSelector);
 
+    ethPeers.setTrailingPeerRequirementsSupplier(synchronizer::calculateTrailingPeerRequirements);
+
+    if (SyncMode.isSnapSync(syncConfig.getSyncMode())
+        || SyncMode.isCheckpointSync(syncConfig.getSyncMode())) {
+      synchronizer.subscribeInSync((b) -> ethPeers.snapServerPeersNeeded(!b));
```
Comment (usmansaleem) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:711]: ` ethPeers.snapServerPeersNeeded` is called twice on L711 and L712, is it by design?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 5]
```diff
@@ -92,12 +102,22 @@ public class EthPeers {
...
+  private final ForkIdManager forkIdManager;
+  private final int snapServerTargetNumber;
+  private final boolean shouldLimitRemoteConnections;
 
   private Comparator<EthPeer> bestPeerComparator;
   private final Bytes localNodeId;
   private RlpxAgent rlpxAgent;
 
   private final Counter connectedPeersCounter;
+  //  private List<ProtocolManager> protocolManagers;
```
Comment (usmansaleem) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:115]: nit: can delete commented code.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 6]
```diff
@@ -411,28 +448,132 @@ public void disconnectWorstUselessPeer() {
...
+          isServer = snapServerChecker.check(peer, peersHeadBlockHeader).get(6L, TimeUnit.SECONDS);
+        } catch (Exception e) {
+          LOG.atTrace()
+              .setMessage("Error checking if peer {} is a snap server. Setting to false.")
+              .addArgument(peer.getLoggableId())
+              .log();
+          peer.setIsServingSnap(false);
+          return;
+        }
+        peer.setIsServingSnap(isServer);
```
Comment (usmansaleem) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:570]: nit: L570 and L571 can be moved inside try block after L561.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java [hunk 7]
```diff
@@ -67,8 +67,19 @@ protected AbstractRetryingPeerTask(
...
+   *
+   * @param peer The peer to assign to the task.
+   * @return True if the peer was assigned, false otherwise.
+   */
+  public boolean assignPeer(final EthPeer peer) {
+    if (isSuitablePeer(peer)) {
+      assignedPeer = Optional.of(peer);
+      return true;
+    } else {
+      return false;
```
Comment (usmansaleem) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingPeerTask.java:81]: should assignedPeer assigned to Optional.empty in else part?

**Review by usmansaleem — COMMENTED** [review]
minor comments otherwise looks good to me.

**Review by usmansaleem — APPROVED** [review]
LGTM

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java (hunk 6 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:570]: I prefer it this way, to make it a bit easier to understand.

**Pull request merged**